engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateBy.java

```diff
@@ -1,96 +1,810 @@
+            modifiedRowSet.remove(downstream.added);
+
+            return downstream;
+        }
+
+        /**
+         * Process the {@link TableUpdate update} provided in the constructor. This performs much work in parallel and
+         * leverages {@link io.deephaven.engine.table.impl.select.analyzers.SelectAndViewAnalyzer.JobScheduler}
+         * extensively
+         */
+        public void processUpdate() {
+            if (redirContext.isRedirected()) {
+                // this call does all the work needed for redirected output sources, for sparse output sources
+                // we will process shifts only after a call to `prepareForParallelPopulation()` on each source
+                redirContext.processUpdateForRedirection(update, source.getRowSet());
+                shiftedRows = RowSetFactory.empty();
             } else {
-                inputSourceSlots[opIdx] = maybeExistingSlot;
+                // for our sparse array output sources, we need to identify which rows will be affected by the upstream
+                // shifts and include them in our parallel update preparations
+                if (update.shifted().nonempty()) {
+                    try (final RowSet prev = source.getRowSet().copyPrev();
+                            final RowSequence.Iterator it = prev.getRowSequenceIterator()) {
+
+                        final RowSetBuilderSequential builder = RowSetFactory.builderSequential();
+                        final int size = update.shifted().size();
+
+                        // get these in ascending order and use a sequential builder
+                        for (int ii = 0; ii < size; ii++) {
+                            final long begin = update.shifted().getBeginRange(ii);
+                            final long end = update.shifted().getEndRange(ii);
+                            final long delta = update.shifted().getShiftDelta(ii);
+
+                            it.advance(begin);
+                            final RowSequence rs = it.getNextRowSequenceThrough(end);
+                            builder.appendRowSequenceWithOffset(rs, delta);
+                        }
+                        shiftedRows = builder.build();
+                    }
+                } else {
+                    shiftedRows = RowSetFactory.empty();
+                }
+            }
+
+            // this is where we leave single-threaded calls and rely on the scheduler to continue the work. Each
+            // call will chain to another until the sequence is complete
+            computeCachedColumnRowsets(
+                    () -> processWindows(
+                            () -> cleanUpAndNotify(
+                                    () -> {
+                                        // signal to the main task that we have completed our work
+                                        waitForResult.complete(null);
+                                    }
+                            )));
```

Comment:
I expected something more iterative, here, as you moved through different windows. Is there something I'm missing?

Reply:
Inside `processWindows()` is another chain of calls serially processing the windows.